scholly/service/core/push.js

```diff
@@ -0,0 +1,186 @@
+    if (err) {
+      winston.error('Failed to remove built directory.');
+    }
+  });
+}
+
+async function makeGHRepo(ghUsername, ghToken) {
+  // check if the repo exists
+  const repoName = ghUsername + '.github.io';
+  const searchQuery = 'q=user:' + ghUsername + ' ' + repoName + ' in:name';
+  const options = {
+    Authorization: 'token ' + ghToken,
+    Accept: 'application/vnd.github.v3+json',
+  };
+  let repoResponse;
+  try {
+    repoResponse = await axios.get(
+      'https://api.github.com/search/repositories?' + searchQuery,
+      {
+        headers: options,
+      }
+    );
+  } catch (err) {
+    console.log(err);
+  }
+
+  if (repoResponse.data.total_count === 1) {
+    console.log('repo exists already');
+  } else {
+    // if it doesn't, make one
+    const createRepoBody = {
+      name: repoName,
+      private: false, // free accounts can only use a public repo for GH pages
+      auto_init: true,
+    };
+    try {
+      await axios({
+        url: 'https://api.github.com/user/repos',
+        method: 'post',
+        headers: options,
+        data: createRepoBody,
+      });
+      console.log('Repo was created successfully');
+    } catch (err) {
+      console.log(err);
+    }
+  }
+
+  // set up pages for the repo regardless of whether its new or not
+  // see https://docs.github.com/en/rest/reference/repos#pages
+  try {
+    const pagesResponse = await axios.get(
+      'https://api.github.com/repos/' + ghUsername + '/' + repoName + '/pages',
+      {
+        headers: {
+          Authorization: 'token ' + ghToken,
+          Accept: 'application/vnd.github.switcheroo-preview+json',
+        },
+      }
+    );
+    console.log('GH Pages already configured for repo');
+  } catch (err) {
+    console.log(err);
+    console.log('GH pages not created yet');
+    createPagesSite(ghUsername, ghToken);
+  }
+}
+
+async function createPagesSite(ghUsername, ghToken) {
+  const pagesBody = {
+    source: {
+      branch: 'main',
+      path: '/',
+    },
+  };
+  const repoName = ghUsername + '.github.io';
+  let createPagesResponse;
+  try {
+    createPagesResponse = await axios({
+      url:
+        'https://api.github.com/repos/' +
+        ghUsername +
+        '/' +
+        repoName +
+        '/pages',
+      method: 'post',
+      headers: {
+        Authorization: 'token ' + ghToken,
+        Accept: 'application/vnd.github.switcheroo-preview+json',
+      },
+      data: pagesBody,
+    });
+    if (createPagesResponse.status === 201) {
+      console.log('Pages site successfully created');
+    } else {
+      // see https://docs.github.com/en/rest/reference/repos#create-a-github-pages-site
+      console.log(createPagesResponse.data);
+    }
+  } catch (err) {
+    console.log(err);
+  }
+}
+
+async function buildPages(ghUsername, ghToken) {
+  const options = {
+    Authorization: 'token ' + ghToken,
+    Accept: 'application/vnd.github.v3+json',
+  };
+  const repoName = ghUsername + '.github.io';
```

Comment:
We use this same code in 3 or 4 places, this may be a bit pedantic 😂, but can we extract this into a seperate function maybe getRepoName(), the reason being is that GitHub Pages allows users to use a seperate name (i.e. not the default repoName.github.io), and if this become a requirement for us, it would be easier to change it in one place :)

Reply:
Yup you're right! I'll pull it up so we only have to change it in one place.